# `Problem Statement: Dynamic Pricing Strategies for Fitness Classes based on Demand, Time, and Location`
Cult fit aims to gain & maintain a competitive advantage by staying ahead of market trends and customer preferences in the fitness industry.

Leveraging a data driven approach gives them valuable insights into customer behavior, demand patterns, and the impact of pricing changes.

In this project, we will help analyze attendance data of various fitness classes and develop a dynamic pricing model that will help Cult fit maximize attendance and increase their revenue per class.

# Step 1: Data Cleaning & Preprocessing
## 1.1 Load Raw Data


In this step, we load the two datasets (April–May and June 2018) 
into pandas DataFrames for further processing.

In [57]:
# Import Required Libraries

import pandas as pd
import numpy as np

# Load datasets

df_apr_may = pd.read_csv("Classes April-May 2018.csv")
df_june = pd.read_csv("Classes June 2018.csv")

# Display shape of datasets

print("April-May Dataset Shape:", df_apr_may.shape)
print("June Dataset Shape:", df_june.shape)



April-May Dataset Shape: (2177, 7)
June Dataset Shape: (1112, 7)


In [58]:
# Preview data
df_apr_may.head()

ActivitySiteID      ActivityDescription  \
0            HXP  20-20-20  2.45pm-3.45pm   
1            HXP  20-20-20  2.45pm-3.45pm   
2            HXP  20-20-20  2.45pm-3.45pm   
3            HXP  20-20-20  2.45pm-3.45pm   
4            HXP  20-20-20  2.45pm-3.45pm   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                               08-Apr-18         14:45:00          25   
1                               15-Apr-18         14:45:00          25   
2                               22-Apr-18         14:45:00          25   
3                               29-Apr-18         14:45:00          25   
4                               06-May-18         14:45:00          25   

   Number Booked  Price (INR)  
0             12        499.0  
1             15        499.0  
2             14        499.0  
3              9        499.0  
4              7        499.0

In [59]:
df_june.head()

ActivitySiteID     ActivityDescription  \
0            BRP  20:20:20  9.30-10.30am   
1            BRP  20:20:20  9.30-10.30am   
2            BRP  20:20:20  9.30-10.30am   
3            BRP  20:20:20  9.30-10.30am   
4            BRP  20:20:20  9.30-10.30am   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                               01-Jun-18          9:30:00          35   
1                               08-Jun-18          9:30:00          35   
2                               15-Jun-18          9:30:00          35   
3                               22-Jun-18          9:30:00          35   
4                               29-Jun-18          9:30:00          35   

   Number Booked  Price (INR)  
0             28          499  
1             35          499  
2             31          499  
3             32          499  
4             19          499

### Inspect & Understand Data Structure

Before merging the datasets, we inspect:

- Column names
- Data types
- Summary statistics
- Missing values

This ensures structural compatibility and identifies the correct key for merging.

In [60]:
# Check column names

print("April-May Columns:\n", df_apr_may.columns)
print("\nJune Columns:\n", df_june.columns)


April-May Columns:
 Index(['ActivitySiteID', 'ActivityDescription',
       'BookingEndDateTime (Month / Day / Year)', 'BookingStartTime',
       'MaxBookees', 'Number Booked', 'Price (INR)'],
      dtype='object')

June Columns:
 Index(['ActivitySiteID', 'ActivityDescription',
       'BookingEndDateTime (Month / Day / Year)', 'BookingStartTime',
       'MaxBookees', 'Number Booked', 'Price (INR)'],
      dtype='object')


In [61]:
# Quick statistical summary
print("April-May Statistical Summary:")
print(df_apr_may.describe())

print("\nJune Statistical Summary:")
print(df_june.describe())



April-May Statistical Summary:
        MaxBookees  Number Booked  Price (INR)
count  2177.000000    2177.000000  2159.000000
mean     32.192007      17.660542  1855.368689
std      14.972919       9.622520   801.420753
min       2.000000       1.000000   499.000000
25%      24.000000      10.000000  1299.000000
50%      30.000000      17.000000  1499.000000
75%      35.000000      24.000000  2499.000000
max      70.000000      64.000000  3999.000000

June Statistical Summary:
        MaxBookees  Number Booked  Price (INR)
count  1112.000000    1112.000000  1112.000000
mean     32.154676      15.969424  1847.561151
std      15.329663       9.169420   775.628834
min       2.000000       1.000000   499.000000
25%      24.000000       9.000000  1299.000000
50%      30.000000      15.000000  1499.000000
75%      35.000000      22.000000  2499.000000
max      70.000000      52.000000  3999.000000


In [62]:
# Check data types

print("April-May Info:\n")
df_apr_may.info()

print("\nJune Info:\n")
df_june.info()


April-May Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2177 entries, 0 to 2176
Data columns (total 7 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   ActivitySiteID                           2177 non-null   object 
 1   ActivityDescription                      2177 non-null   object 
 2   BookingEndDateTime (Month / Day / Year)  2177 non-null   object 
 3   BookingStartTime                         2177 non-null   object 
 4   MaxBookees                               2177 non-null   int64  
 5   Number Booked                            2177 non-null   int64  
 6   Price (INR)                              2159 non-null   float64
dtypes: float64(1), int64(2), object(4)
memory usage: 119.2+ KB

June Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1112 entries, 0 to 1111
Data columns (total 7 columns):
 #   Column                                   Non-Null Count  Dt

In [63]:
# Check missing values in April-May dataset
print("Missing Values - April-May")
df_apr_may.isnull().sum()


Missing Values - April-May


ActivitySiteID                              0
ActivityDescription                         0
BookingEndDateTime (Month / Day / Year)     0
BookingStartTime                            0
MaxBookees                                  0
Number Booked                               0
Price (INR)                                18
dtype: int64

In [64]:
# Check missing values in June dataset
print("Missing Values - June")
df_june.isnull().sum()


Missing Values - June


ActivitySiteID                             0
ActivityDescription                        0
BookingEndDateTime (Month / Day / Year)    0
BookingStartTime                           0
MaxBookees                                 0
Number Booked                              0
Price (INR)                                0
dtype: int64

## 1.2 Clean Data:
### Handle Missing Values

Missing Values Identified:

• There are total 18 rows missing (column :Price (INR)) in April-May dataset

Since Price is critical for dynamic pricing and revenue modeling,
we will impute missing values using median price.

Median is chosen because it is robust to outliers.



In [65]:
# Check median price
median_price = df_apr_may['Price (INR)'].median()
print("Median Price:", median_price)


Median Price: 1499.0


In [66]:
# Fill missing prices with median

df_apr_may['Price (INR)'] = df_apr_may['Price (INR)'].fillna(median_price)

# Verify missing values
print("Remaining Missing in Price:",
      df_apr_may['Price (INR)'].isnull().sum())


Remaining Missing in Price: 0


### Remove Duplicates

Before merging datasets, we check and remove duplicate rows 
in both April-May and June datasets.

Duplicate records may distort:
- Demand estimation
- Revenue calculations
- Pricing strategy

Hence, duplicates must be removed.



In [67]:
# Check duplicates in April-May dataset
print("April-May Duplicate Rows:", df_apr_may.duplicated().sum())

# Check duplicates in June dataset
print("June Duplicate Rows:", df_june.duplicated().sum())


April-May Duplicate Rows: 0
June Duplicate Rows: 0


### Duplicate Check Result

• April–May Dataset: 0 duplicate rows  
• June Dataset: 0 duplicate rows  

This confirms:

- No data duplication present  
- No risk of inflated attendance or revenue figures  
- Dataset integrity maintained  

Hence, no duplicate removal was required.


### Correct Data Types using astype()

To ensure consistency before merging, we explicitly convert columns 
to appropriate data types using df.astype().

Correct data types ensure:

* Accurate merging
* Proper numerical analysis
 * Reliable forecasting model performance


In [68]:
# Check data types

print("April-May data types:\n")
print(df_apr_may.dtypes)

print("\nJune data types:\n")
df_june.dtypes


April-May data types:

ActivitySiteID                              object
ActivityDescription                         object
BookingEndDateTime (Month / Day / Year)     object
BookingStartTime                            object
MaxBookees                                   int64
Number Booked                                int64
Price (INR)                                float64
dtype: object

June data types:



ActivitySiteID                             object
ActivityDescription                        object
BookingEndDateTime (Month / Day / Year)    object
BookingStartTime                           object
MaxBookees                                  int64
Number Booked                               int64
Price (INR)                                 int64
dtype: object

After inspecting both datasets, the following inconsistency was identified:

### Data Type Inconsistency

- `Price (INR)` had different data types:
  - April–May → float64
  - June → int64

Since price is a numerical variable used for revenue calculation 
and regression modeling, it must have a consistent data type across datasets.

---

### Action Taken

To ensure structural consistency before merging and modeling:

- Converted `Price (INR)` to float64 in both datasets using `astype()`.

Note:
Date and time format standardization will be handled separately 
under the "Transform Data" section as per project instructions.


In [69]:
# Standardize Price column type

df_apr_may['Price (INR)'] = df_apr_may['Price (INR)'].astype(float)
df_june['Price (INR)'] = df_june['Price (INR)'].astype(float)

## Integrate Data

Both April–May and June datasets:

- Contain identical schema (7 columns)
- Represent different time periods
- Do not share relational keys

Since all columns are identical in both datasets,
we perform an outer merge on all columns`(no common identifiers (e.g., customer IDs, class IDs) present in data set)`

This effectively integrates both time periods
into a single consolidated dataset.

In [70]:
# Merge on all columns (since schema is identical)

df_merged = pd.merge(
    df_apr_may,
    df_june,
    how='outer'
)

print("Merged Dataset Shape:", df_merged.shape)

Merged Dataset Shape: (3289, 7)


In [71]:
print("April-May rows:", len(df_apr_may))
print("June rows:", len(df_june))
print("Merged rows:", len(df_merged))


April-May rows: 2177
June rows: 1112
Merged rows: 3289


## 1.3 Transform Data
The following transformations are applied:



In [72]:
import re

# 1. Standardize Date Format (Specifying the exact format to avoid warnings)
df_merged['BookingEndDateTime (Month / Day / Year)'] = pd.to_datetime(
    df_merged['BookingEndDateTime (Month / Day / Year)'], 
    format='%d-%b-%y'
)
# 2. Splitting Logic (Handles "20-20-20" and "Zumba9.30am")
def robust_split(description):
    if pd.isna(description):
        return None, None
    
    desc_str = str(description).strip()
    
    # --- SPECIAL CASE: 20-20-20 or 20:20:20 ---
    # These names start with numbers, so we handle them first
    if desc_str.startswith('20:20:20') or desc_str.startswith('20-20-20'):
        name = "20-20-20"
        # The name is 8 characters long, the rest is the time slot
        time = desc_str[8:].strip()
        return name, time
    
    # --- GENERAL CASE: Everything else (e.g., Zumba9.30am) ---
    # We find the position of the very first number (0-9)
    match = re.search(r'\d', desc_str)
    if match:
        idx = match.start()
        # Everything before the first digit is the Name
        name = desc_str[:idx].strip("- ")
        # Everything from the first digit onwards is the Time
        time = desc_str[idx:].strip()
        return name, time
    
    # If no number is found at all, the whole thing is the Name
    return desc_str, None

# 3. Apply the Split and Final Formatting
df_merged[['ActivityName', 'TimeSlot']] = df_merged['ActivityDescription'].apply(lambda x: pd.Series(robust_split(x)))

# Standardize Site IDs and casing
df_merged['ActivityName'] = df_merged['ActivityName'].str.title()
df_merged['ActivitySiteID'] = df_merged['ActivitySiteID'].str.upper()
df_merged['TimeSlot'] = df_merged['TimeSlot'].str.lower().str.replace(" ", "")

# 4. Feature Engineering: Demand (Occupancy Rate)
# We use round(..., 2) to keep the numbers clean for analysis
df_merged['OccupancyRate'] = round(df_merged['Number Booked'] / df_merged['MaxBookees'], 2)

# 5. Final Column Selection
cols_to_keep = [
    'ActivitySiteID', 'ActivityName', 
    'BookingEndDateTime (Month / Day / Year)', 'BookingStartTime','MaxBookees',
    'Number Booked', 'OccupancyRate', 'Price (INR)'
]
df_final = df_merged[cols_to_keep]

print("Successfully Completed with Robust Splitting!")
display(df_final.head(5))

Successfully Completed with Robust Splitting!


ActivitySiteID ActivityName BookingEndDateTime (Month / Day / Year)  \
0            BRP     20-20-20                              2018-05-01   
1            BRP     20-20-20                              2018-04-03   
2            BRP     20-20-20                              2018-06-05   
3            BRP     20-20-20                              2018-05-08   
4            BRP     20-20-20                              2018-04-10   

  BookingStartTime  MaxBookees  Number Booked  OccupancyRate  Price (INR)  
0         10:00:00          35             22           0.63        499.0  
1         10:00:00          35             32           0.91        499.0  
2         10:00:00          35             25           0.71        499.0  
3         10:00:00          35             30           0.86        499.0  
4         10:00:00          35             26           0.74        499.0

In [73]:
# Check if columns have empty strings or just spaces
empty_strings = (df_final['ActivityName'] == "").sum()
only_spaces = df_final['ActivityName'].str.isspace().sum()

print(f"Empty strings found: {empty_strings}")
print(f"Rows that are just blank spaces: {only_spaces}")

# FIX: Convert blanks/spaces to actual NaN so Python recognizes them as missing
import numpy as np
df_final = df_final.replace(r'^\s*$', np.nan, regex=True)

print("\nMissing values after exposing hidden blanks:")
print(df_final.isnull().sum())

Empty strings found: 39
Rows that are just blank spaces: 0

Missing values after exposing hidden blanks:
ActivitySiteID                              0
ActivityName                               39
BookingEndDateTime (Month / Day / Year)     0
BookingStartTime                            0
MaxBookees                                  0
Number Booked                               0
OccupancyRate                               0
Price (INR)                                 0
dtype: int64


In [74]:
df_final['ActivityName'] = df_final['ActivityName'].fillna("Unknown Activity")
#Final Verification
print("Final Missing Value Check:")
print(df_final.isnull().sum())

Final Missing Value Check:
ActivitySiteID                             0
ActivityName                               0
BookingEndDateTime (Month / Day / Year)    0
BookingStartTime                           0
MaxBookees                                 0
Number Booked                              0
OccupancyRate                              0
Price (INR)                                0
dtype: int64


### Feature Scaling Strategy

Numerical features may require normalization
or scaling before applying regression models.

The following numerical columns will be considered:

- MaxBookees
- Number Booked
- Price (INR)

However, scaling will be applied during the modeling phase
after performing train-test split.

This ensures:

- No data leakage
- Consistent feature distribution
- Proper regression coefficient interpretation


## 2.2 Ensure Data Consistency

After merging both datasets:

- Verified total row count matches expected total.
- Confirmed no records were lost during integration.
- Ensured consistent data types across all columns.
- Confirmed no unexpected duplicate column creation.

The merged dataset structure is consistent and reliable.


In [75]:
# Check structure
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3289 entries, 0 to 3288
Data columns (total 8 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   ActivitySiteID                           3289 non-null   object        
 1   ActivityName                             3289 non-null   object        
 2   BookingEndDateTime (Month / Day / Year)  3289 non-null   datetime64[ns]
 3   BookingStartTime                         3289 non-null   object        
 4   MaxBookees                               3289 non-null   int64         
 5   Number Booked                            3289 non-null   int64         
 6   OccupancyRate                            3289 non-null   float64       
 7   Price (INR)                              3289 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(3)
memory usage: 205.7+ KB


## 2.3 Conduct Quality Checks

After merging the datasets, quality validation is performed to ensure:

- No missing values exist after integration
- No duplicate records were introduced
- No negative price values are present
- No overbooking cases exist (Number Booked > MaxBookees)

These checks ensure data reliability before moving to analysis and modeling.


In [76]:
# Create a dictionary to store our 'Quality Findings'
quality_issues = {}

# Check 1: Zero or Negative Prices
zero_prices = df_final[df_final['Price (INR)'] <= 0]
quality_issues['Zero/Negative Prices'] = len(zero_prices)

# Check 2: Zero Capacity Classes
zero_capacity = df_final[df_final['MaxBookees'] <= 0]
quality_issues['Zero Capacity Errors'] = len(zero_capacity)

# Check 3: Overbooking (Occupancy > 100%)
overbooked = df_final[df_final['OccupancyRate'] > 1.0]
quality_issues['Overbooked Sessions'] = len(overbooked)

# Check 4: Missing Activity Names
missing_names = df_final['ActivityName'].isnull().sum()
quality_issues['Missing Activity Names'] = missing_names

print("Quality Checks Completed.")

Quality Checks Completed.


In [77]:
print("======= DATA QUALITY REPORT =======")
print(f"Total Records Analyzed: {len(df_final)}")
print("-" * 35)

for issue, count in quality_issues.items():
    status = " ISSUE FOUND" if count > 0 else " CLEAN"
    print(f"{issue:25}: {count:<5} | {status}")

print("-" * 35)
print("Recommendation: " + ("Clean data before proceeding." if sum(quality_issues.values()) > 0 else "Data is consistent and ready for EDA."))

======= DATA QUALITY REPORT =======
Total Records Analyzed: 3289
-----------------------------------
Zero/Negative Prices     : 0     |  CLEAN
Zero Capacity Errors     : 0     |  CLEAN
Overbooked Sessions      : 18    |  ISSUE FOUND
Missing Activity Names   : 0     |  CLEAN
-----------------------------------
Recommendation: Clean data before proceeding.


## 2.4 Quality Report Summary

After conducting post-merge validation checks:

### Findings

- Total Rows: 3289
- Missing Values: 0
- Duplicate Rows: 0
- Negative Prices: 0
- Overbooked Cases: 18

### Observation

18 records were identified where:

Number Booked > MaxBookees

This represents overbooking and is considered a data anomaly.

Such inconsistencies may distort demand modeling and utilization analysis.


#### Handling Overbooking Cases

Since overbooked cases (18 records) represent data inconsistencies,
these rows are removed to maintain data integrity.

This ensures:

- Accurate demand modeling
- Correct utilization calculation
- Reliable dynamic pricing strategy



In [78]:
# 1. Cap OccupancyRate at 1.0
# We use .loc to safely update the values without getting warnings
df_final.loc[df_final['OccupancyRate'] > 1.0, 'OccupancyRate'] = 1.0

# 2. Verify the fix
remaining_issues = (df_final['OccupancyRate'] > 1.0).sum()

print(f"Final cleanup complete. Remaining overbooked sessions: {remaining_issues}")
print("Data is now 100% consistent.")



Final cleanup complete. Remaining overbooked sessions: 0
Data is now 100% consistent.


## Cleaned and Preprocessed Dataset Submission

The final cleaned dataset includes:

- Standardized data types
- Proper date and time formatting
- No missing values
- No duplicate records
- No negative price values
- Treatment of overbooking anomalies

The dataset has been exported as:

**Cleaned_Preprocessed_Fitness_Data.csv**

This dataset is now ready for exploratory analysis, 
demand forecasting, and dynamic pricing model development.


In [79]:
# Export cleaned and preprocessed dataset
df_final.to_csv("Cleaned_Preprocessed_Fitness_Data.csv", index=False)
print("Cleaned dataset exported successfully.")

Cleaned dataset exported successfully.


## Final Conclusion – Data Cleaning, Integration & Quality Assurance

The datasets were thoroughly cleaned by handling missing values, 
standardizing data types, and formatting date/time columns appropriately.  
April–May and June datasets were successfully merged using `pd.merge()` 
without data loss. Post-merge validation confirmed no missing values, 
duplicates, or negative prices, and overbooking anomalies were removed.  

The final dataset is clean, consistent, and ready for exploratory analysis and modeling.
